Add documentation for window function avoidance workarounds (#16064)
MaterializeInc/materialize: doc/user/content/sql/patterns/window-functions.md

---
title: "Window functions (OVER clause)"
description: "Avoid window functions, e.g., use lateral joins as a streaming-efficient stand-in for many window functions."
menu:
  main:
    parent: 'sql-patterns'
---

> Materialize currently does not handle window functions (`OVER` clause) efficiently for large partition sizes: when an input record is added/removed/changed, the system recomputes the results for the entire partition that the changed record belongs to. We are planning to make window functions more efficient in Materialize, but for the time being, this page shows some workarounds for avoiding the use of window functions.

Reviewer: ```suggestion
Some query patterns that are commonly used in batch are tricky to support efficiently in streaming due to the unbounded nature of the input data. Window functions, such as `LAG`, `LEAD`, `ROW_NUMBER`, `FIRST_VALUE`, are one of those patterns.

As in, Materialize doesn't handle window functions efficiently for large partition sizes: when an input record is added, removed, or changed, the system recomputes the results for the entire partition that the changed record belongs to. For this reason, we recommend using the alternative approaches in this page whenever possible, until support for window functions in Materialize is refactored to provide better performance.
```

> Materialize currently does not handle window functions (`OVER` clause) efficiently for large partition sizes: when an input record is added/removed/changed, the system recomputes the results for the entire partition that the changed record belongs to. We are actively working on making window functions more efficient in Materialize, but for the time being, this page shows some workarounds for avoiding the use of window functions.

Reviewer 2: I would leave out "We're actively working on" and replace by a less committal statement. We're planning it at the moment.

Author: Changed to "We are planning to make window functions more efficient in Materialize"

> Note that this page **does _not_ refer to temporal windows**, i.e., when a window belongs to a time interval. Examples of temporal windows are when we select only the last hour of an infinite stream of records, or when we group input records by which hour they occurred in. Such queries are efficiently supported in Materialize by using [temporal filters and temporal groupings](/sql/patterns/temporal-filters).

Author: I mostly applied this, but also changed the first part of the next paragraph to make it absolutely clear that this document is not about temporal windows. Could you please check, @morsapaes?

> By "window functions", this page means functions such as `LAG`, `LEAD`, `ROW_NUMBER`, `FIRST_VALUE`. You can access these functions with the `OVER` clause, in which you use the `PARTITION BY` clause to group input records into partitions by a key, and use the `ORDER BY` clause to order records inside each partition. Window functions compute results based on the position of records inside these partitions. For example, `x - LAG(x)` computes the differences of attribute `x` between each pair of _consecutive_ rows in each partition.

Reviewer: ```suggestion
```

Author: You mean we should delete this paragraph?

## Example data

> We use the following table as input for later examples queries on this page:

Reviewer: ```suggestion
Let's use the following sample data as input for examples:
```

```sql
CREATE TABLE cities (
    name text NOT NULL,
    state text NOT NULL,
    pop int NOT NULL
);

INSERT INTO cities VALUES
    ('Los_Angeles', 'CA', 3979576),
    ('Phoenix', 'AZ', 1680992),
    ('Houston', 'TX', 2320268),
    ('San_Diego', 'CA', 1423851),
    ('San_Francisco', 'CA', 881549),
    ('New_York', 'NY', 8336817),
    ('Dallas', 'TX', 1343573),
    ('San_Antonio', 'TX', 1547253),
    ('San_Jose', 'CA', 1021795),
    ('Chicago', 'IL', 2695598),
    ('Austin', 'TX', 978908);
```

## Top K using `ROW_NUMBER`

In other databases, a popular way of computing the top _K_ records per key is to use the `ROW_NUMBER` window function. For example, to get the 3 most populous city in each state:
```
SELECT state, name
FROM (
  SELECT state, name, ROW_NUMBER() OVER
    (PARTITION BY state ORDER BY pop) as row_num
  FROM cities
)
WHERE row_num <= 3;
```

> If there are states that have many cities, then it is better for performance to express this query using a lateral join (or `DISTINCT ON` is _K_ = 1) instead of window functions, which we explain in [Top K by group](/sql/patterns/top-k).

Reviewer: ```suggestion
If there are states that have many cities, a more performant way to express this in Materialize is to use a lateral join (or `DISTINCT ON`, if _K_ = 1) instead of window functions. See [Top K by group](/sql/patterns/top-k) for the alternative implementation.
```

Reviewer: We should provide the alternative here, rather than point at the Top K page. Or, maybe we should merge the two pages? Just thinking out loud, no need to block this PR on that.

## `FIRST_VALUE`/`LAST_VALUE` of an entire partition

> Suppose that we want to compute the ratio of each city's population vs. the most populous city in the same state. We can do so using window functions as follows:

Reviewer: ```suggestion
Suppose that you want to compute the ratio of each city's population vs. the most populous city in the same state. You can do so using window functions as follows:
```

```
SELECT state, name,
       CAST(pop AS float) / FIRST_VALUE(pop)
         OVER (PARTITION BY state ORDER BY pop DESC)
FROM cities;
```

> For better performance, we can rewrite this query as follows: We first compute the largest population of each state by a simple grouped aggregation, and then join against that:

Reviewer: ```suggestion
For better performance, you can rewrite this query to first compute the largest population of each state using an aggregation, and then join against that:
```

```
SELECT cities.state, name, CAST(pop as float) / max_pops.max_pop
FROM cities,

>      (SELECT state, MAX(pop) as max_pop

Author: Well, it's true that `MAX` is much less efficient and needs more memory than e.g., `SUM`. But if the user really has to calculate the max, then I don't know any better thing to do than just calling `MAX`. (Btw. if the optimizer know that the input is monotonic, then `MAX` is fine.)

>      (SELECT state, MAX(pop) as max_pop

Author: What do you mean by "there is no aggregation"?

>      (SELECT state, MAX(pop) as max_pop

Reviewer: What I meant was if you'd want e.g. the maximum population overall, without grouping by state:

```sql
SELECT MAX(pop) as max_pop
FROM cities;
```

I was just trying to understand our stance IRWT `MAX()`, though, so SGTM and thank you for the explanation!

      FROM cities
      GROUP BY state) max_pops
WHERE cities.state = max_pops.state;
```

If the `ROW_NUMBER` would be called with an expression that is different from the expression in the `ORDER BY` clause, then the inner query would have to be replaced by a [`DISTINCT ON` query](/sql/patterns/top-k).

## Aggregate window functions over an entire partition

> Suppose that we want to compute the ratio of each city's population vs. the total population of the city's state. We can do so using an aggregate window function as follows:

Reviewer: ```suggestion
Suppose that you want to compute the ratio of each city's population vs. the total population of the city's state. You can do so using an aggregate window function as follows:
```

```
SELECT state, name,
       CAST(pop AS float) / SUM(pop)
         OVER (PARTITION BY state)
FROM cities;
```

> Aggregate window functions are not yet supported in Materialize, but we can rewrite this query as follows: We first compute the total population of each state by a simple grouped aggregation, and then join against that:

Reviewer: ```suggestion
Aggregate window functions are not yet supported in Materialize, but you can rewrite this query to first compute the total population of each state using an aggregation, and then join against that:
```

```
SELECT cities.state, name, CAST(pop as float) / total_pops.total_pop
FROM cities,
     (SELECT state, SUM(pop) as total_pop
      FROM cities
      GROUP BY state) total_pops
WHERE cities.state = total_pops.state;
```

## `LAG`/`LEAD` for time series

> If the input has a column that advances by regular amounts, then `LAG` and `LEAD` can be replaced by an equi-join where we perform some arithmetic on the key to related consecutive elements. Suppose that we have the following data:

Reviewer: ```suggestion
If the input has a column that advances by regular amounts, then `LAG` and `LEAD` can be replaced by an equi-join. Suppose that you have the following data:
```

```
CREATE TABLE measurements(time timestamp, value float);
INSERT INTO measurements VALUES
    (TIMESTAMP '2007-02-01 15:04:01', 8),
    (TIMESTAMP '2007-02-01 15:05:01', 9),
    (TIMESTAMP '2007-02-01 15:06:01', 12);
```

> We can compute the differences between consecutive measurements using `LAG()`:

Reviewer: ```suggestion
You can compute the differences between consecutive measurements using `LAG()`:
```
